glean-core/android/src/main/java/mozilla/telemetry/glean/upload/Upload.kt

```diff
@@ -0,0 +1,105 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.telemetry.glean.upload
+
+import org.json.JSONObject
+import com.sun.jna.Structure
+import com.sun.jna.Pointer
+import com.sun.jna.Union
+import mozilla.telemetry.glean.Glean
+import mozilla.telemetry.glean.net.HeadersList
+import mozilla.telemetry.glean.rust.getRustString
+
+@Suppress("ClassNaming")
+@Structure.FieldOrder("tag", "uuid", "path", "body", "headers")
+internal class Upload_Body(
+    // NOTE: We need to provide defaults here, so that JNA can create this object.
+    @JvmField val tag: Byte? = null,
+    @JvmField val uuid: Pointer? = null,
+    @JvmField val path: Pointer? = null,
+    @JvmField val body: Pointer? = null,
+    @JvmField val headers: Pointer? = null
+) : Structure() {
+    fun toPingRequest(): PingRequest {
+        return PingRequest(
+            this.path!!.getRustString(),
+            this.body!!.getRustString(),
+            this.headers!!.getRustString()
+        )
+    }
+}
+
+@Structure.FieldOrder("tag", "upload")
+internal open class FfiPingUploadTask(
+    // NOTE: We need to provide defaults here, so that JNA can create this object.
+    @JvmField var tag: Byte = 2,
+    @JvmField var upload: Upload_Body = Upload_Body()
+) : Union() {
+    class ByValue : FfiPingUploadTask(), Structure.ByValue
+
+    fun toPingUploadTask(): PingUploadTask {
+        return when (this.tag.toInt()) {
+            0 -> PingUploadTask.Wait
+            1 -> {
+                this.readField("upload")
+                PingUploadTask.Upload(this.upload.toPingRequest())
+            }
+            else -> PingUploadTask.Done
+        }
+    }
+}
+
+/**
+ * The PingUploadTask makes it easier to consume Upload_Body.
+ */
+internal class PingRequest(
+    val path: String,
+    val body: String,
+    headers: String
+) {
+    val headers: HeadersList = headersFromJSONString(headers)
+
+    companion object {
+        fun headersFromJSONString(str: String): HeadersList {
+            val jsonHeaders = JSONObject(str)
+            val headers: MutableList<Pair<String, String>> = mutableListOf()
+            for (key in jsonHeaders.keys()) {
+                headers.add(Pair(key, jsonHeaders.get(key).toString()))
+            }
+
+            Glean.configuration.pingTag?.let {
+                headers.add(Pair("X-Debug-ID", it))
+            }
+
+            return headers
+        }
+    }
+}
+
+/**
+ * The PingUploadTask makes it easier to consume FfiPingUploadTask.
+ */
+internal sealed class PingUploadTask {
+    /**
+     * A PingRequest popped from the front of the queue.
+     */
+    class Upload(val request: PingRequest) : PingUploadTask()
+
+    /**
+     * A flag signaling that the pending pings directories are not done being processed,
+     * thus the requester should wait and come back later.
+     */
+    object Wait : PingUploadTask()
```

Comment:
Linter complains about this being a class simply because it doesn't really need to be one, since it doesn't have any properties it can be a static object. I could suppress that, but I don't see the point.

> the compiler complains about equality here

Hum, I didn't see that when I was writing this, only the linter error really 🤔

Reply:
The linter is complaining about equality :D It says you either need to provide an implementation for `equals` or make it an object.

Reply:
TIL!